lib/alexa_request_verifier/verifier.rb

```diff
@@ -4,6 +4,121 @@
 module AlexaRequestVerifier
   # A namespace for all of our verifiers to live under
   # @since 0.1
-  module Verifier
+  class Verifier
+    attr_accessor :configuration
+
+    # Create a new AlexaRequestVerifier::Verifier object
+    #
+    # @yield the configuration block
+    # @yieldparam config [AlexaRequestVerifier::Configuration] the configuration object
+    def initialize
+      @configuration = AlexaRequestVerifier::Configuration.new
+
+      yield @configuration if block_given?
+    end
+
+    # Validate a request object from Sinatra.
+    # Raise an error if it is not valid.
+    #
+    # @param [Sinatra::Request] request a Sinatra HTTP Request
+    #
+    # @raise [AlexaRequestVerifier::InvalidCertificateURIError]
+    #   there was a problem validating the certificate URI from your request
+    #
+    # @return [nil] will always return nil
+    def valid!(request)
+      signature_certificate_url = request.env['HTTP_SIGNATURECERTCHAINURL']
+
+      AlexaRequestVerifier::Verifier::CertificateURIVerifier.valid!(signature_certificate_url) if @configuration.verify_uri?
+
+      raw_body = request.body.read
+      request.body.rewind
+
+      check_that_request_is_timely(raw_body) if @configuration.verify_timeliness?
+
+      check_that_request_is_valid(signature_certificate_url, request, raw_body)
+
+      true
+    end
+
+    # Validate a request object from Sinatra.
+    # Return a boolean.
+    #
+    # @param [Sinatra::Request] request a Sinatra HTTP Request
+    # @return [Boolean] is the request valid?
+    def valid?(request)
+      begin
+        valid!(request)
+      rescue AlexaRequestVerifier::BaseError => e
+        puts e
+
+        return false
+      end
+
+      true
+    end
+
+    # Used to configure AlexaRequestVerifier.
+    #
+    # @example
+    #    AlexaRequestVerifier.configure do |c|
+    #      c.some_config_option = true
+    #    end
+    #
+    # @yield the configuration block
+    # @yieldparam config [AlexaRequestVerifier::Configuration] the configuration object
+    def configure
+      yield @configuration
+    end
+
+    private
+
+    # Prevent replays of requests by checking that they are timely.
+    #
+    # @param [String] raw_body the raw body of our https request
+    # @raise [AlexaRequestVerifier::InvalidRequestError] raised when the timestamp is not timely, or is not set
+    def check_that_request_is_timely(raw_body)
+      request_json = JSON.parse(raw_body)
+
+      raise AlexaRequestVerifier::InvalidRequestError, 'Timestamp field not present in request' if request_json.fetch('request', {}).fetch('timestamp', nil).nil?
+
+      raise AlexaRequestVerifier::InvalidRequestError, 'Request is from more than 150 seconds ago' unless Time.parse(request_json['request']['timestamp'].to_s) >= (Time.now - REQUEST_THRESHOLD)
```

Comment:
should this be

```ruby
      raise AlexaRequestVerifier::InvalidRequestError, "Request is from more than #{REQUEST_THRESHOLD} seconds ago" unless Time.parse(request_json['request']['timestamp'].to_s) >= (Time.now - REQUEST_THRESHOLD)
```

?

No strong opinions, but I find longer lines like this easier to read when they're in standard conditional block, eg:

```ruby
if Time.parse(request_json['request']['timestamp'].to_s) < (Time.now - REQUEST_THRESHOLD)
  raise AlexaRequestVerifier::InvalidRequestError, "Request is from more than #{REQUEST_THRESHOLD} seconds ago"
end
```

less horizontal scrolling :)

Reply:
For both the first point and the comment below about `REQUEST_THRESHOLD` being configurable - more than happy to update the message - that's best practice for sure - `REQUEST_THRESHOLD` itself if actually defined in the specification from Amazon as being a maximum of 150 seconds between the timestamp being generated and our application processing that request. See the second paragraph here: https://developer.amazon.com/docs/custom-skills/host-a-custom-skill-as-a-web-service.html#timestamp

---

For your second point about the conditional - I've tried wherever possible to follow the Ruby style guide, verifying the gem's code with Rubocop at every point. The style guide says that these guard statements should be expressed as a single line. What I can do to make it more readable would be something like this:

```ruby
request_is_timely = (Time.parse(request_json['request']['timestamp'].to_s) < (Time.now - REQUEST_THRESHOLD))
raise AlexaRequestVerifier::InvalidRequestError, "Request is from more than #{REQUEST_THRESHOLD} seconds ago" unless request_is_timely
```

You can see the 'cop' in question here: http://www.rubydoc.info/github/bbatsov/rubocop/Rubocop/Cop/Style/GuardClause

Would the above be agree-able?